### PR TITLE
List Subcommand (Implementation)

### PR DESCRIPTION
**Pull request by carolynzech**
Implementation of the list subcommand (and updates to the RFC).

As a larger test, I ran on the standard library (`kani list -Z list -Z function-contracts -Z mem-predicates ./library --std`) and manually verified that the results were correct. I pasted the output below.

Contracts:
|       | Function                                         | Contract Harnesses (#[kani::proof_for_contract])       |
| ----- | ------------------------------------------------ | ------------------------------------------------------ |
|       | alloc::layout::Layout::from_size_align_unchecked | alloc::layout::verify::check_from_size_align_unchecked |
|       | ascii::ascii_char::AsciiChar::from_u8            | ascii::ascii_char::verify::check_from_u8               |
|       | ascii::ascii_char::AsciiChar::from_u8_unchecked  | ascii::ascii_char::verify::check_from_u8_unchecked     |
|       | char::convert::from_u32_unchecked                | char::convert::verify::check_from_u32_unchecked        |
|       | char::methods::verify::as_ascii_clone            | char::methods::verify::check_as_ascii_ascii_char       |
|       |                                                  | char::methods::verify::check_as_ascii_non_ascii_char   |
|       | intrinsics::typed_swap                           | intrinsics::verify::check_typed_swap_u8                |
|       |                                                  | intrinsics::verify::check_typed_swap_char              |
|       |                                                  | intrinsics::verify::check_typed_swap_non_zero          |
|       | mem::swap                                        | mem::verify::check_swap_primitive                      |
|       |                                                  | mem::verify::check_swap_adt_no_drop                    |
|       | ptr::align_offset                                | ptr::verify::check_align_offset_zst                    |
|       |                                                  | ptr::verify::check_align_offset_u8                     |
|       |                                                  | ptr::verify::check_align_offset_u16                    |
|       |                                                  | ptr::verify::check_align_offset_u32                    |
|       |                                                  | ptr::verify::check_align_offset_u64                    |
|       |                                                  | ptr::verify::check_align_offset_u128                   |
|       |                                                  | ptr::verify::check_align_offset_4096                   |
|       |                                                  | ptr::verify::check_align_offset_5                      |
|       | ptr::alignment::Alignment::as_nonzero            | ptr::alignment::verify::check_as_nonzero               |
|       | ptr::alignment::Alignment::as_usize              | ptr::alignment::verify::check_as_usize                 |
|       | ptr::alignment::Alignment::log2                  | ptr::alignment::verify::check_log2                     |
|       | ptr::alignment::Alignment::mask                  | ptr::alignment::verify::check_mask                     |
|       | ptr::alignment::Alignment::new                   | ptr::alignment::verify::check_new                      |
|       | ptr::alignment::Alignment::new_unchecked         | ptr::alignment::verify::check_new_unchecked            |
|       | ptr::alignment::Alignment::of                    | ptr::alignment::verify::check_of_i32                   |
|       | ptr::non_null::NonNull::<T>::new                 | ptr::non_null::verify::non_null_check_new              |
|       | ptr::non_null::NonNull::<T>::new_unchecked       | ptr::non_null::verify::non_null_check_new_unchecked    |
|       | ptr::read_volatile                               | ptr::verify::check_read_u128                           |
|       | ptr::unique::Unique::<T>::as_non_null_ptr        | ptr::unique::verify::check_as_non_null_ptr             |
|       | ptr::unique::Unique::<T>::as_ptr                 | ptr::unique::verify::check_as_ptr                      |
|       | ptr::unique::Unique::<T>::new                    | ptr::unique::verify::check_new                         |
|       | ptr::unique::Unique::<T>::new_unchecked          | ptr::unique::verify::check_new_unchecked               |
|       | ptr::verify::mod_inv_copy                        | ptr::verify::check_mod_inv                             |
|       | ptr::write_volatile                              | NONE                                                   |
| Total | 24                                               | 34                                                     |

Standard Harnesses (#[kani::proof]):
1. ptr::unique::verify::check_as_mut
2. ptr::unique::verify::check_as_ref
3. ptr::unique::verify::check_cast

Terminal view (`--pretty` format):
list

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
I would prefer if the compiler is not aware that this is a list command.

File: kani-compiler/src/args.rs
Comment (carolynzech) [reply kani-compiler/src/args.rs:50]: Per slack, no -- will fix. Going to move to draft until then, since this will take me a bit of time to do and I don't want anyone wasting time reviewing in the meantime.

File: kani-compiler/src/kani_middle/attributes.rs
Comment (carolynzech) [reply kani-compiler/src/kani_middle/attributes.rs:548]: People may want to write harnesses before they write the contracts. For instance, if I'm planning to write a contract for a function, I could write:

```rust
fn foo(x: usize) -> usize { ... }

#[kani::proof_for_contract(foo)
fn harness() {
  let x: usize = kani::any();
  foo(x)
}
```

It makes sense to error during verification, since we can't verify a contract that doesn't exist. But I didn't think it made sense to punish people for being halfway done; they may want to use the list command to see what work they have left to do.

That being said, I'm not even sure I could make this distinction if we're making the compiler unaware of the list command, so we may have to make it a compiler error.

**Comment by celinval** [comment]
What is `# of contracts`? Why does `mem::swap` has 0?

**Comment by carolynzech** [comment]
> What is `# of contracts`? Why does `mem::swap` has 0?

# of contracts is the number of contracts applied to the function. See the PR description for the explanation for `mem::swap`--it's because we don't currently count modifies clauses as contracts.

File: rfc/src/rfcs/0013-list.md
Comment (celinval) [inline rfc/src/rfcs/0013-list.md:61]: Why a numbered list?

File: rfc/src/rfcs/0013-list.md
Comment (celinval) [inline rfc/src/rfcs/0013-list.md:158]: This is per function definition not instantiation, right?

File: rfc/src/rfcs/0013-list.md
Comment (celinval) [inline rfc/src/rfcs/0013-list.md:156]: TBH, I don't think you need to add this level of detail. I think saying high-level that the metadata will now include a list of functions that contains contract and information about their harnesses.

File: kani-compiler/src/kani_middle/metadata.rs
Comment (celinval) [inline kani-compiler/src/kani_middle/metadata.rs:59]:
```suggestion
        let fn_name = item.name(def_id);
```

**Review by celinval — COMMENTED** [review]
Thanks for doing this!! Per our offline discussion, I think we should remove the `# of Contracts`. As a bonus it should simplify the compiler changes.

File: rfc/src/rfcs/0013-list.md
Comment (carolynzech) [reply rfc/src/rfcs/0013-list.md:61]: Because it's a concise way to see how many standard harnesses you have--we can report the list and the total simultaneously.

File: rfc/src/rfcs/0013-list.md
Comment (carolynzech) [reply rfc/src/rfcs/0013-list.md:158]: Yes.

**Comment by carolynzech** [comment]
@celinval Thanks for the feedback -- I removed contracts count and simplified the RFC. Also per offline discussion, I stabilized the `gen_contracts_metadata` function. It still uses unstable `DefId`s, since in this line:

```rust
else if let Some((target_name, target_def_id, _)) = attributes.interpret_for_contract_attribute()
```

`target_def_id` is an unstable `DefId` and AFAIK there's no way to convert an unstable `DefId` to a stable one. LMK if I'm mistaken and I'll fix it.

**Comment by celinval** [comment]
> @celinval Thanks for the feedback -- I removed contracts count and simplified the RFC. Also per offline discussion, I stabilized the `gen_contracts_metadata` function. It still uses unstable `DefId`s, since in this line:
> 
> ```rust
> else if let Some((target_name, target_def_id, _)) = attributes.interpret_for_contract_attribute()
> ```
> 
> `target_def_id` is an unstable `DefId` and AFAIK there's no way to convert an unstable `DefId` to a stable one. LMK if I'm mistaken and I'll fix it.

You could use this function to get the FnDef: https://github.com/model-checking/kani/blob/68387e4cf462541028892772d2fc12da8abd6a6c/kani-compiler/src/kani_middle/codegen_units.rs#L118

File: rfc/src/rfcs/0013-list.md
Comment (celinval) [reply rfc/src/rfcs/0013-list.md:61]: I personally like the previous format better with the explicit total. But I'll leave it up to you. Thanks

**Review by feliperodri — CHANGES_REQUESTED** [review]
Thanks so much for this new feature! Only a few more minor comments to go before merging it.

File: kani-driver/src/list/collect_metadata.rs
Comment (carolynzech) [reply kani-driver/src/list/collect_metadata.rs:27]: I use ordered maps and sets because the data should be in alphabetical order. It would be weird if you ran list twice on the same inputs and your output was printed in a different order.

Comment (carolynzech) [reply kani-driver/src/list/collect_metadata.rs:27]: See comment on line 21:

```rust
 // We use ordered maps and sets so that the output is in lexographic order (and consistent across invocations).
```

File: tests/script-based-pre/kani_list/list.sh
Comment (carolynzech) [reply tests/script-based-pre/kani_list/list.sh:7]: They're actually slightly different scripts -- one invokes `kani` and other invokes `cargo kani`. We have separate tests for each because `cargo kani` creates multiple `KaniMetadata` objects, and we want to test that we correctly combine the information from multiple metadata objects.

File: kani-compiler/src/kani_middle/metadata.rs
Comment (carolynzech) [reply kani-compiler/src/kani_middle/metadata.rs:54]: I think I'll leave it as-is because in this case:

```rust
if let Some(cf) = fn_to_data.get_mut(&target_def_id) {
    cf.harnesses.push(function);
}
```

nothing gets inserted into the hashmap, and I think `fold` makes more sense when every iteration folds something into the accumulator.

File: kani-driver/src/list/output.rs
Comment (carolynzech) [reply kani-driver/src/list/output.rs:57]: I'm not quite sure what you mean--I could write syntactically different code to achieve this same goal, but in any case it would still require iterating through each `ContractedFunction` to find the maximum widths.
I agree that this code is a bit long. I took inspiration for this format from [Adrian's `kani-cov` work](https://github.com/model-checking/kani/pull/3121). I can see a future where we have a new crate in `kani-driver` that just prints Markdown tables given inputs, which would abstract away some of this detail.

File: rfc/src/rfcs/0013-list.md
Comment (camshaft) [reply rfc/src/rfcs/0013-list.md:195]: I'm not sure I follow the comment here. Bolero harnesses use the `kani::proof` attribute just like any other proof. Is the aim to distinguish bolero harnesses from kani harnesses?

Comment (camshaft) [reply rfc/src/rfcs/0013-list.md:195]: Ah I just expanded the paragraph above and that's exactly what the doc says:

> It would be nice to differentiate between regular Kani harnesses and Bolero harnesses. Bolero harnesses invoke Kani using conditional compilation

I'm open to any suggestions on how to improve this.

**Review by zhassan-aws — COMMENTED** [review]
Nice work!

File: rfc/src/rfcs/0013-list.md
Comment (carolynzech) [reply rfc/src/rfcs/0013-list.md:32]: Yes, my bad -- forgot to update this when we removed the contracts count.

File: kani-driver/src/list/output.rs
Comment (carolynzech) [reply kani-driver/src/list/output.rs:24]: Earlier commits for this PR actually used a crate like that. I decided to change it to this because I thought it would be really nice to have it closer to Markdown. That way, when we're sharing the table format to track our verify std goal, we can format Markdown tables in documents instead of having screenshots of huge tables.
Since the code for this earlier version is committed, it would be really easy for me to change it back. I think for now we can see how much we actually take advantage of it being Markdown-adjacent and go from there?

File: rfc/src/rfcs/0013-list.md
Comment (carolynzech) [inline rfc/src/rfcs/0013-list.md:32]:
```suggestion
Each row of the "Contracts" table consists of a function under contract and its contract harnesses.
```

File: rfc/src/rfcs/0013-list.md
Comment (carolynzech) [inline rfc/src/rfcs/0013-list.md:33]:
```suggestion
```

**Review by zhassan-aws — APPROVED** [review]
Thanks!

File: rfc/src/rfcs/0013-list.md
Comment (celinval) [reply rfc/src/rfcs/0013-list.md:194]: In theory, I think we can look at the list of functions reachable from the harness and determine if it uses bolero. I'm not sure the extra overhead is worth it though.

Comment (camshaft) [reply rfc/src/rfcs/0013-list.md:194]: Yeah I'm not sure it's really worth distinguishing a bolero harness vs a vanilla Kani one... It's not super clear to me what one would actually do with that information.

**Review by feliperodri — APPROVED** [review]

**Pull request merged**
